utils/openqa_investigator.py

```diff
@@ -0,0 +1,564 @@
+
+    filtered_results = []
+
+    for test_failure in test_failures:
+        if test_failure.error_message and \
+            re.search(error_pattern, test_failure.error_message):
+            filtered_results.append(test_failure)
+
+    filtered_job = deepcopy(job)
+    filtered_job.failures[job.get_job_name()] = filtered_results
+
+    return filtered_job
+
+def group_by_error(test):
+    if not test.error_message:
+        return "no error printed\n(probably a native openQA test)"
+
+    desc_lines = test.error_message.split("\n")
+
+    result = ""
+    max_chars = 40
+
+    # attempt to find the line with the relevant result
+    for line in reversed(desc_lines):
+        if any(map(lambda error: error in line, IGNORED_ERRORS)): # ignore certain
+            continue
+        elif line == "" or re.search("^\s+$", line): # whitespace
+            continue
+        else:
+            if result: # last two lines
+                result = line[:max_chars] + ".*\n" + result
+                break
+            else:
+                result = line[:max_chars] + ".*"
+
+    if result == "":
+        return "ignored error"
+
+    return result
+
+def group_by_template(test):
+    # obtain template name according to construction format of
+    # https://github.com/QubesOS/qubes-core-admin/blob/f60334/qubes/tests/__init__.py#L1352
+    # Will catch most common tests.
+    template = test.name.split("_")[-1]
+    template = template.split("-pool")[0] # remove trailing "-pool"
+
+    if re.search(r"^[a-z\-]+\-\d+(\-xfce)?$", template): # [template]-[ver]
+        return template
+    else:
+        msg  = "Test's name '{}' doesn't specify a template.\n".format(test.name)
+        msg += "  The test suite may not include template information in the"
+        msg += " test's name."
+        logging.warning(msg)
+
+        return "unspecifed template"
+
+def plot_group_by_test(title, jobs, test_suite, outfile=None):
+    y_fn = lambda test: test.title
+    plot_simple(title, jobs, test_suite, y_fn, outfile)
+
+def plot_group_by_template(title, jobs, test_suite, outfile=None):
+    plot_simple(title, jobs, test_suite, group_by_template, outfile)
+
+    def group_by_template(test):
+        # obtain template name according to construction format of
+        # https://github.com/QubesOS/qubes-core-admin/blob/f60334/qubes/tests/__init__.py#L1352
+        # Will catch most common tests.
+        template = test.name.split("_")[-1]
+        template = template.split("-pool")[0] # remove trailing "-pool"
+
+        if re.search(r"^[a-z\-]+\-\d+(\-xfce)?$", template): # [template]-[ver]
+            return template
+
+        msg  = "Test's name '{}' doesn't specify a template.\n".format(test.name)
+        msg += "  The test suite '{}' may not include".format(test_suite)
+        msg += "template information in the test's name."
+        raise Exception(msg)
+
+    plot_simple(title, jobs, test_suite, group_by_template, outfile)
```

Comment:
This looks like a rebase error. `group_by_template` is defined twice, and `plot_simple` is also called twice.